### ISSUE TITLE
Mounting using AutoFS doesn't work

### ISSUE THREAD

**Issue by ScuttleSE** [rclone/rclone#3246]
#### Output of `rclone version`
rclone v1.47.0
- os/arch: linux/amd64
- go version: go1.12.3

#### Describe the issue
I have been trying to add a few rclone-mounts to my autofs configuration without success

Mounting the drive using the exact commandline that autofs uses works though, so something is weird...

```handle_packet: type = 3
handle_packet_missing_indirect: token 20, name TeamDrive-Archive, request pid 3291
attempting to mount entry /media/GoogleTest/TeamDrive-Archive
lookup_mount: lookup(file): looking up TeamDrive-Archive
lookup_mount: lookup(file): TeamDrive-Archive -> -fstype=fuse,config=/home/scuttle/.config/rclone/rclone.conf	:rclonefs\#dom-td01\:
parse_mount: parse(sun): expanded entry: -fstype=fuse,config=/home/scuttle/.config/rclone/rclone.conf	:rclonefs\#dom-td01\:
parse_mount: parse(sun): gathered options: fstype=fuse,config=/home/scuttle/.config/rclone/rclone.conf
parse_mount: parse(sun): dequote(":rclonefs\#dom-td01\:") -> :rclonefs#dom-td01:
parse_mount: parse(sun): core of entry: options=fstype=fuse,config=/home/scuttle/.config/rclone/rclone.conf, loc=:rclonefs#dom-td01:
sun_mount: parse(sun): mounting root /media/GoogleTest, mountpoint TeamDrive-Archive, what rclonefs#dom-td01:, fstype fuse, options config=/home/scuttle/.config/rclone/rclone.conf
do_mount: rclonefs#dom-td01: /media/GoogleTest/TeamDrive-Archive type fuse options config=/home/scuttle/.config/rclone/rclone.conf using module generic
mount_mount: mount(generic): calling mkdir_path /media/GoogleTest/TeamDrive-Archive
mount_mount: mount(generic): calling mount -t fuse -o config=/home/scuttle/.config/rclone/rclone.conf rclonefs#dom-td01: /media/GoogleTest/TeamDrive-Archive
```

Just running "mount -t fuse -o config=/home/scuttle/.config/rclone/rclone.conf rclonefs#dom-td01: /media/GoogleTest/TeamDrive-Archive" mounts the folder just fine, but when automount tries it, it just hangs on that last line

**Comment by ScuttleSE** [comment]
Oh, and I'm using the mount helper script from here https://github.com/ncw/rclone/wiki/rclone-fstab-mount-helper-script

**Comment by ncw** [comment]
I suggest you add the `--log-file` option or the `--syslog` option to the rclone invocation so you can see any errors and post them here - thanks!

**Comment by ScuttleSE** [comment]
This is what the rclone-log showed:

2019/06/07 06:54:10 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vvvv" "--log-file=/tmp/autofs.rclone" "mount" "--config=/home/scuttle/.config/rclone/rclone.conf" "--allow-other" "--default-permissions" "--max-read-ahead=16M" "--drive-impersonate=redacted@example.com" "dom-01:" "/media/GoogleTest/redacted@example.com"]
2019/06/07 06:54:10 DEBUG : Using config file from "/home/scuttle/.config/rclone/rclone.conf"
2019/06/07 06:54:10 DEBUG : Google drive root '': Mounting on "/media/GoogleTest/redacted@example.com"
2019/06/07 06:54:11 DEBUG : Adding path "vfs/forget" to remote control registry
2019/06/07 06:54:11 DEBUG : Adding path "vfs/refresh" to remote control registry
2019/06/07 06:54:11 DEBUG : Adding path "vfs/poll-interval" to remote control registry
2019/06/07 06:54:11 DEBUG : : Root: 
2019/06/07 06:54:11 DEBUG : : >Root: node=/, err=<nil>
```

**Comment by ScuttleSE** [comment]
Tried adding --daemon to the commandline as suggested here: https://forum.rclone.org/t/issues-with-rclone-mount-root-node-err-nil/6749

Still not working, but I got another set of messages in the log

```
2019/06/07 09:46:13 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vvvv" "--log-file=/tmp/autofs.rclone" "mount" "--config=/home/scuttle/.config/rclone/rclone.conf" "--allow-other" "--daemon" "--default-permissions" "--max-read-ahead=16M" "--drive-impersonate=redacted@example.com" "dom-01:" "/media/GoogleTest/redacted@example.com"]
2019/06/07 09:46:13 DEBUG : Using config file from "/home/scuttle/.config/rclone/rclone.conf"
2019/06/07 09:46:13 DEBUG : rclone: Version "v1.47.0" finishing with parameters ["rclone" "-vvvv" "--log-file=/tmp/autofs.rclone" "mount" "--config=/home/scuttle/.config/rclone/rclone.conf" "--allow-other" "--daemon" "--default-permissions" "--max-read-ahead=16M" "--drive-impersonate=redacted@example.com" "dom-01:" "/media/GoogleTest/redacted@example.com"]
2019/06/07 09:46:13 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vvvv" "--log-file=/tmp/autofs.rclone" "mount" "--config=/home/scuttle/.config/rclone/rclone.conf" "--allow-other" "--daemon" "--default-permissions" "--max-read-ahead=16M" "--drive-impersonate=redacted@example.com" "dom-01:" "/media/GoogleTest/redacted@example.com"]
2019/06/07 09:46:13 DEBUG : Using config file from "/home/scuttle/.config/rclone/rclone.conf"
```

**Comment by ncw** [comment]
There are no obvious errors in there... The first one looks like it is working.

So the symptoms are that
- you `ls` the automounted directory
- rclone appears to have mounted the directory (can you check in another terminal)
- however your `ls` hangs due to some automounter interaction?

So do we think what is wrong is some interaction between automount and rclone?

Adding `--daemon` was a good idea.

**Comment by ScuttleSE** [comment]
Yes

- autofs catches me trying to ls the folder /media/GoogleTest/redacted@example.com
- autofs checks what to do in the file /etc/autofs/auto.google where I have this line:
`redacted@example.com	-fstype=fuse,config=/home/scuttle/.config/rclone/rclone.conf,allow-other,daemon,default-permissions,max-read-ahead=16M,drive-impersonate=redacted@example.com	:rclonefs#dom-01:`
- autofs tries to mount using the rclonefs mount-helper, the script at https://github.com/ncw/rclone/wiki/rclone-fstab-mount-helper-script
- Nothing seems to happen

The folder never gets mounted, but if I run the commandline from the rclone-log it mounts it just fine.

**Comment by ncw** [comment]
> The folder never gets mounted,

Can you check from another terminal that isn't hung to see if the folder did get mounted or not?

**Comment by ScuttleSE** [comment]
Same as the first terminal. the ls-command just hangs

No additional lines in the rclone logfile

**Comment by ncw** [comment]
If you look at `mount` or `cat /proc/mounts` is it mounted in there?

**Comment by ScuttleSE** [comment]
Nope, doesn't show up in either. 
Other autofs mounts do show up.

**Comment by ScuttleSE** [comment]
Here's the rclone.conf btw

```
[dom-01]
type = drive
client_id =
client_secret =
scope = drive
root_folder_id =
service_account_file = /home/scuttle/.config/rclone/dom/rclone-01.json
```

**Comment by ncw** [comment]
What you could try is when it has locked up, find the rclone PID then `kill -QUIT pid`.

This will cause rclone to emit a stack trace of exactly what it is up to to the log file.  Hopefully that will give us a clue!

**Comment by ScuttleSE** [comment]
Hm, `kill -QUIT` didn't kill the process so I had some issues getting it... but this ended up in the logfile after a few tries

```SIGQUIT: quit
PC=0x55dc0f5528d1 m=7 sigcode=0

goroutine 0 [idle]:
runtime.futex(0xc000128148, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7f9282ba0d20, 0x55dc0f4fe075, ...)
	/usr/lib/go/src/runtime/sys_linux_amd64.s:535 +0x21
runtime.futexsleep(0xc000128148, 0x7f9200000000, 0xffffffffffffffff)
	/usr/lib/go/src/runtime/os_linux.go:46 +0x4b
runtime.notesleep(0xc000128148)
	/usr/lib/go/src/runtime/lock_futex.go:151 +0xa5
runtime.stopm()
	/usr/lib/go/src/runtime/proc.go:1936 +0xc5
runtime.findrunnable(0xc000040500, 0x0)
	/usr/lib/go/src/runtime/proc.go:2399 +0x550
runtime.schedule()
	/usr/lib/go/src/runtime/proc.go:2525 +0x220
runtime.park_m(0xc000001c80)
	/usr/lib/go/src/runtime/proc.go:2605 +0xa7
runtime.mcall(0x800000)
	/usr/lib/go/src/runtime/asm_amd64.s:299 +0x53

goroutine 1 [syscall]:
syscall.Syscall6(0x101, 0xffffffffffffff9c, 0xc000293770, 0x80000, 0x0, 0x0, 0x0, 0x7ffe8109cd8f, 0xc000136a80, 0xc0001b6380)
	/usr/lib/go/src/syscall/asm_linux_amd64.s:44 +0x5
syscall.openat(0xffffffffffffff9c, 0x7ffe8109cd97, 0x26, 0x80000, 0xc000000000, 0x12, 0x55dc10157bed, 0x12)
	/usr/lib/go/src/syscall/zsyscall_linux_amd64.go:68 +0xc2
syscall.Open(...)
	/usr/lib/go/src/syscall/syscall_linux.go:138
os.openFileNolog(0x7ffe8109cd97, 0x26, 0x0, 0x55dc00000000, 0xc000391cb8, 0x4e445cd8, 0xff438e3aedf6b969)
	/usr/lib/go/src/os/file_unix.go:201 +0x93
os.OpenFile(0x7ffe8109cd97, 0x26, 0x0, 0x55dc00000000, 0x5, 0x55dc10147527, 0xc0001e6950)
	/usr/lib/go/src/os/file.go:284 +0x61
os.Open(...)
	/usr/lib/go/src/os/file.go:265
github.com/ncw/rclone/cmd/mountlib.checkMountEmpty(0x7ffe8109cd97, 0x26, 0x0, 0x0)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mountlib/mount.go:43 +0x61
github.com/ncw/rclone/cmd/mountlib.NewMountCommand.func1(0xc0003b0280, 0xc000422dc0, 0x2, 0xa)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mountlib/mount.go:263 +0x452
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).execute(0xc0003b0280, 0xc000422c80, 0xa, 0xa, 0xc0003b0280, 0xc000422c80)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:766 +0x2b0
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x55dc1138e840, 0xc000391f88, 0x55dc0f4f9021, 0xc0000a0058)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:852 +0x2ee
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:800
github.com/ncw/rclone/cmd.Main()
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/cmd.go:488 +0x49
main.main()
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/rclone.go:13 +0x22

goroutine 5 [syscall]:
os/signal.signal_recv(0x0)
	/usr/lib/go/src/runtime/sigqueue.go:139 +0x9e
os/signal.loop()
	/usr/lib/go/src/os/signal/signal_unix.go:23 +0x24
created by os/signal.init.0
	/usr/lib/go/src/os/signal/signal_unix.go:29 +0x43

rax    0xca
rbx    0xc000128000
rcx    0x55dc0f5528d3
rdx    0x0
rdi    0xc000128148
rsi    0x80
rbp    0x7f9282ba0ce8
rsp    0x7f9282ba0ca0
r8     0x0
r9     0x0
r10    0x0
r11    0x286
r12    0x5
r13    0x7f927bbe2a7f
r14    0x7f9282ba1700
r15    0x0
rip    0x55dc0f5528d1
rflags 0x286
cs     0x33
fs     0x0
gs     0x0
```

**Comment by ncw** [comment]
Ah, that makes sense.  rclone is blocked on `github.com/ncw/rclone/cmd/mountlib.checkMountEmpty` - it is listing the mountpoint to see if it is empty.  This is likely to cause some recursion or other funkiness with autofs.

If you add the ` --allow-non-empty` flag it will skip that check - perhaps that will make it work!

**Comment by ncw** [comment]
Assuming that does work, I wonder if there is some way rclone could detect it is being run from autofs (maybe an environment variable) and turn that check off automatically?

**Comment by ScuttleSE** [comment]
Hm, it still hangs, but it seems to get a bit further, I get a fusemount-process

```root     23678 23665  0 19:33 ?        00:00:00 fusermount -o allow_other,default_permissions,subtype=rclone,fsname=dom-01:,nonempty -- /media/GoogleTest/redacted@example.com

**Comment by ScuttleSE** [comment]
Trace from rclone this time

```SIGQUIT: quit
PC=0x56131bbe1a80 m=0 sigcode=0

goroutine 0 [idle]:
runtime.epollwait(0x5, 0x7ffdedcd6e90, 0xffffffff00000080, 0x24000000000, 0xffffffff000eb3eb, 0x7ffd00000000, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/lib/go/src/runtime/sys_linux_amd64.s:675 +0x20
runtime.netpoll(0xc000057901, 0xc000057901)
	/usr/lib/go/src/runtime/netpoll_epoll.go:71 +0x128
runtime.findrunnable(0xc000040500, 0x0)
	/usr/lib/go/src/runtime/proc.go:2380 +0x525
runtime.schedule()
	/usr/lib/go/src/runtime/proc.go:2525 +0x220
runtime.park_m(0xc000095980)
	/usr/lib/go/src/runtime/proc.go:2605 +0xa7
runtime.mcall(0x56131c7d3f70)
	/usr/lib/go/src/runtime/asm_amd64.s:299 +0x53

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc000431348)
	/usr/lib/go/src/runtime/sema.go:56 +0x3b
sync.(*WaitGroup).Wait(0xc000431340)
	/usr/lib/go/src/sync/waitgroup.go:130 +0x67
github.com/ncw/rclone/vendor/bazil.org/fuse.mount(0x7ffdedcd8d97, 0x26, 0xc000438690, 0xc0000a0fc0, 0xc000341328, 0x0, 0x0, 0x0)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount_linux.go:100 +0x95d
github.com/ncw/rclone/vendor/bazil.org/fuse.Mount(0x7ffdedcd8d97, 0x26, 0xc00042c8c0, 0x8, 0x8, 0x7ffdedcd8d8f, 0x0, 0xc000431300)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fuse.go:171 +0x169
github.com/ncw/rclone/cmd/mount.mount(0x56131d033140, 0xc000095680, 0x7ffdedcd8d97, 0x26, 0x56131bc36670, 0x56131bc344ed, 0x21, 0x3, 0x0)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mount/mount.go:87 +0x169
github.com/ncw/rclone/cmd/mount.Mount(0x56131d033140, 0xc000095680, 0x7ffdedcd8d97, 0x26, 0x6, 0x1)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mount/mount.go:132 +0x9b
github.com/ncw/rclone/cmd/mountlib.NewMountCommand.func1(0xc0002d5400, 0xc000337550, 0x2, 0xb)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mountlib/mount.go:286 +0x28f
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).execute(0xc0002d5400, 0xc000337340, 0xb, 0xb, 0xc0002d5400, 0xc000337340)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:766 +0x2b0
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x56131da1d840, 0xc000365f88, 0x56131bb88021, 0xc0000a0058)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:852 +0x2ee
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:800
github.com/ncw/rclone/cmd.Main()
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/cmd.go:488 +0x49
main.main()
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/rclone.go:13 +0x22

goroutine 19 [syscall]:
os/signal.signal_recv(0x0)
	/usr/lib/go/src/runtime/sigqueue.go:139 +0x9e
os/signal.loop()
	/usr/lib/go/src/os/signal/signal_unix.go:23 +0x24
created by os/signal.init.0
	/usr/lib/go/src/os/signal/signal_unix.go:29 +0x43

goroutine 21 [IO wait]:
internal/poll.runtime_pollWait(0x7f096213be38, 0x72, 0xffffffffffffffff)
	/usr/lib/go/src/runtime/netpoll.go:182 +0x58
internal/poll.(*pollDesc).wait(0xc000341458, 0x72, 0x1001, 0x1000, 0xffffffffffffffff)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:87 +0x9d
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc000341440, 0xc00044e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/internal/poll/fd_unix.go:169 +0x19d
os.(*File).read(...)
	/usr/lib/go/src/os/file_unix.go:263
os.(*File).Read(0xc000259170, 0xc00044e000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/os/file.go:108 +0x72
bufio.(*Scanner).Scan(0xc000454f20, 0x56131d014de0)
	/usr/lib/go/src/bufio/scan.go:213 +0xa6
github.com/ncw/rclone/vendor/bazil.org/fuse.lineLogger(0xc000431340, 0x56131c7e88f2, 0x13, 0x56131cfd9028, 0x56131d014de0, 0xc000259170)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount.go:28 +0xfc
created by github.com/ncw/rclone/vendor/bazil.org/fuse.mount
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount_linux.go:98 +0x8dd

goroutine 22 [IO wait]:
internal/poll.runtime_pollWait(0x7f096213bc98, 0x72, 0xffffffffffffffff)
	/usr/lib/go/src/runtime/netpoll.go:182 +0x58
internal/poll.(*pollDesc).wait(0xc000341518, 0x72, 0x1001, 0x1000, 0xffffffffffffffff)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:87 +0x9d
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc000341500, 0xc00044c000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/internal/poll/fd_unix.go:169 +0x19d
os.(*File).read(...)
	/usr/lib/go/src/os/file_unix.go:263
os.(*File).Read(0xc000259180, 0xc00044c000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/os/file.go:108 +0x72
bufio.(*Scanner).Scan(0xc00006cf20, 0x56131d014de0)
	/usr/lib/go/src/bufio/scan.go:213 +0xa6
github.com/ncw/rclone/vendor/bazil.org/fuse.lineLogger(0xc000431340, 0x56131c7e6f71, 0x12, 0xc000343950, 0x56131d014de0, 0xc000259180)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount.go:28 +0xfc
created by github.com/ncw/rclone/vendor/bazil.org/fuse.mount
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount_linux.go:99 +0x94c

rax    0xfffffffffffffffc
rbx    0xffffffff
rcx    0x56131bbe1a80
rdx    0x80
rdi    0x5
rsi    0x7ffdedcd6e90
rbp    0x7ffdedcd7490
rsp    0x7ffdedcd6e50
r8     0x0
r9     0x3
r10    0xffffffff
r11    0x202
r12    0x2
r13    0x56131da38260
r14    0x0
r15    0x0
rip    0x56131bbe1a80
rflags 0x202
cs     0x33
fs     0x0
gs     0x0
```

**Comment by ncw** [comment]
That actually looks like it is working...  Did you try with the `--daemon` flag as well?

**Comment by ScuttleSE** [comment]
Still behaving the same way, just hanging there when trying to list the folder

```2019/06/09 13:58:11 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "--daemon" "-vvvv" "--log-file=/tmp/autofs.rclone" "mount" "--config=/home/scuttle/.config/rclone/rclone.conf" "--allow-other" "--daemon" "--default-permissions" "--allow-non-empty" "--max-read-ahead=16M" "--drive-impersonate=redacted@example.com" "dom-01:" "/media/GoogleTest/redacted@example.com"]
2019/06/09 13:58:11 DEBUG : Using config file from "/home/scuttle/.config/rclone/rclone.conf"
2019/06/09 13:58:11 DEBUG : rclone: Version "v1.47.0" finishing with parameters ["rclone" "--daemon" "-vvvv" "--log-file=/tmp/autofs.rclone" "mount" "--config=/home/scuttle/.config/rclone/rclone.conf" "--allow-other" "--daemon" "--default-permissions" "--allow-non-empty" "--max-read-ahead=16M" "--drive-impersonate=redacted@example.com" "dom-01:" "/media/GoogleTest/redacted@example.com"]
2019/06/09 13:58:11 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "--daemon" "-vvvv" "--log-file=/tmp/autofs.rclone" "mount" "--config=/home/scuttle/.config/rclone/rclone.conf" "--allow-other" "--daemon" "--default-permissions" "--allow-non-empty" "--max-read-ahead=16M" "--drive-impersonate=redacted@example.com" "dom-01:" "/media/GoogleTest/redacted@example.com"]
2019/06/09 13:58:11 DEBUG : Using config file from "/home/scuttle/.config/rclone/rclone.conf"
2019/06/09 13:58:11 DEBUG : Google drive root '': Mounting on "/media/GoogleTest/redacted@example.com"
SIGQUIT: quit
PC=0x5630a5634a80 m=0 sigcode=0

goroutine 0 [idle]:
runtime.epollwait(0x7ffe00000005, 0x7ffe4910d2c0, 0xffffffff00000080, 0x24000000001, 0xffffffff000eb3eb, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/lib/go/src/runtime/sys_linux_amd64.s:675 +0x20
runtime.netpoll(0xc000057901, 0x1)
	/usr/lib/go/src/runtime/netpoll_epoll.go:71 +0x128
runtime.findrunnable(0xc000042a00, 0x0)
	/usr/lib/go/src/runtime/proc.go:2380 +0x525
runtime.schedule()
	/usr/lib/go/src/runtime/proc.go:2525 +0x220
runtime.park_m(0xc00012cd80)
	/usr/lib/go/src/runtime/proc.go:2605 +0xa7
runtime.mcall(0x5630a6226f70)
	/usr/lib/go/src/runtime/asm_amd64.s:299 +0x53

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc00046abe8)
	/usr/lib/go/src/runtime/sema.go:56 +0x3b
sync.(*WaitGroup).Wait(0xc00046abe0)
	/usr/lib/go/src/sync/waitgroup.go:130 +0x67
github.com/ncw/rclone/vendor/bazil.org/fuse.mount(0x7ffe4910ed97, 0x26, 0xc0004093b0, 0xc0001f6e40, 0xc0003aaf08, 0x0, 0x0, 0x0)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount_linux.go:100 +0x95d
github.com/ncw/rclone/vendor/bazil.org/fuse.Mount(0x7ffe4910ed97, 0x26, 0xc0004663c0, 0x8, 0x8, 0x7ffe4910ed8f, 0x0, 0xc00046aba0)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/fuse.go:171 +0x169
github.com/ncw/rclone/cmd/mount.mount(0x5630a6a86140, 0xc00012cc00, 0x7ffe4910ed97, 0x26, 0x5630a5689670, 0x5630a56874ed, 0x21, 0x3, 0x0)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mount/mount.go:87 +0x169
github.com/ncw/rclone/cmd/mount.Mount(0x5630a6a86140, 0xc00012cc00, 0x7ffe4910ed97, 0x26, 0x6, 0x1)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mount/mount.go:132 +0x9b
github.com/ncw/rclone/cmd/mountlib.NewMountCommand.func1(0xc0003b6280, 0xc0001626c0, 0x2, 0xc)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/mountlib/mount.go:286 +0x28f
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).execute(0xc0003b6280, 0xc000162600, 0xc, 0xc, 0xc0003b6280, 0xc000162600)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:766 +0x2b0
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x5630a7470840, 0xc000397f88, 0x5630a55db021, 0xc000092058)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:852 +0x2ee
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:800
github.com/ncw/rclone/cmd.Main()
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/cmd/cmd.go:488 +0x49
main.main()
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/rclone.go:13 +0x22

goroutine 5 [syscall]:
os/signal.signal_recv(0x0)
	/usr/lib/go/src/runtime/sigqueue.go:139 +0x9e
os/signal.loop()
	/usr/lib/go/src/os/signal/signal_unix.go:23 +0x24
created by os/signal.init.0
	/usr/lib/go/src/os/signal/signal_unix.go:29 +0x43

goroutine 18 [IO wait]:
internal/poll.runtime_pollWait(0x7fba2c2a2dd8, 0x72, 0xffffffffffffffff)
	/usr/lib/go/src/runtime/netpoll.go:182 +0x58
internal/poll.(*pollDesc).wait(0xc0003ab038, 0x72, 0x1001, 0x1000, 0xffffffffffffffff)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:87 +0x9d
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc0003ab020, 0xc00047d000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/internal/poll/fd_unix.go:169 +0x19d
os.(*File).read(...)
	/usr/lib/go/src/os/file_unix.go:263
os.(*File).Read(0xc000285138, 0xc00047d000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/os/file.go:108 +0x72
bufio.(*Scanner).Scan(0xc00005a720, 0x5630a6a67de0)
	/usr/lib/go/src/bufio/scan.go:213 +0xa6
github.com/ncw/rclone/vendor/bazil.org/fuse.lineLogger(0xc00046abe0, 0x5630a623b8f2, 0x13, 0x5630a6a2c028, 0x5630a6a67de0, 0xc000285138)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount.go:28 +0xfc
created by github.com/ncw/rclone/vendor/bazil.org/fuse.mount
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount_linux.go:98 +0x8dd

goroutine 19 [IO wait]:
internal/poll.runtime_pollWait(0x7fba2c2a2c38, 0x72, 0xffffffffffffffff)
	/usr/lib/go/src/runtime/netpoll.go:182 +0x58
internal/poll.(*pollDesc).wait(0xc0003ab0f8, 0x72, 0x1001, 0x1000, 0xffffffffffffffff)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:87 +0x9d
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc0003ab0e0, 0xc00047c000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/internal/poll/fd_unix.go:169 +0x19d
os.(*File).read(...)
	/usr/lib/go/src/os/file_unix.go:263
os.(*File).Read(0xc000285148, 0xc00047c000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go/src/os/file.go:108 +0x72
bufio.(*Scanner).Scan(0xc00006cf20, 0x5630a6a67de0)
	/usr/lib/go/src/bufio/scan.go:213 +0xa6
github.com/ncw/rclone/vendor/bazil.org/fuse.lineLogger(0xc00046abe0, 0x5630a6239f71, 0x12, 0xc0003fec50, 0x5630a6a67de0, 0xc000285148)
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount.go:28 +0xfc
created by github.com/ncw/rclone/vendor/bazil.org/fuse.mount
	/build/rclone/src/rclone-1.47.0/.gopath/src/github.com/ncw/rclone/vendor/bazil.org/fuse/mount_linux.go:99 +0x94c

rax    0xfffffffffffffffc
rbx    0xffffffff
rcx    0x5630a5634a80
rdx    0x80
rdi    0x5
rsi    0x7ffe4910d2c0
rbp    0x7ffe4910d8c0
rsp    0x7ffe4910d280
r8     0x0
r9     0x0
r10    0xffffffff
r11    0x202
r12    0xffffffffffffffff
r13    0x2
r14    0x1
r15    0x2
rip    0x5630a5634a80
rflags 0x202
cs     0x33
fs     0x0
gs     0x0
```

**Comment by ncw** [comment]
> Still behaving the same way, just hanging there when trying to list the folder

I misread the traceback, you are right, rclone is waiting for `fusermount` to complete.

Can you `strace` the fusermount process - that will give us an idea what it is doing (if anything).

Another idea, in the original rclone invocation add `</dev/null >/dev/null 2>/dev/null` just to make sure that rclone isn't trying to use any stdio

**Comment by ScuttleSE** [comment]
Hm... how would I strace the fusermount-process? I take it autofs launches it when it tries to mount

**Comment by ScuttleSE** [comment]
Huh, that's interesting, adding `</dev/null >/dev/null 2>/dev/null` seems to have fixed it

**Comment by ncw** [comment]
> Hm... how would I strace the fusermount-process? I take it autofs launches it when it tries to mount

With `strace -p PID` where PID is the fusermount PID



> Huh, that's interesting, adding `</dev/null >/dev/null 2>/dev/null` to the mount-helper seems to have fixed it

Ah!  That means it was blocked on writing to stdout /stderr or reading from stdin which is probably closed when running under automount.

So is it working properly now?

Using `--daemon` should be doing the same thing thing I think so that is a bug...

**Comment by ScuttleSE** [comment]
Yes, but it seems like I need to remove `--daemon`
After that it seems to work fine

It does seem to mount the drive multiple times though...

Right now I have three mounts of the same remote to the same local folder...

**Comment by ScuttleSE** [comment]
Wait, no...restarted, and I keep getting `Too many levels of symbolic links`when listing mounts

**Comment by ncw** [comment]
> Never mind, typo in a thing here

Did you get it working with `--daemon`?

**Comment by ScuttleSE** [comment]
Ah, nope, not with `daemon`, only with `</dev/null >/dev/null 2>/dev/null` in the mount helper

**Comment by ncw** [comment]
> Ah, nope, not with `daemon`, only with `</dev/null >/dev/null 2>/dev/null` in the mount helper

OK so that plus `--allow-non-empty` gets rclone working with autofs?

Can you send a diff of the working mount helper script (or just paste what you have now)?

**Comment by ScuttleSE** [comment]
```#!/bin/bash
remote=$1
mountpoint=$2
shift 2

# Process -o parameters
while getopts :o: opts; do
    case $opts in
        o)
            params=${OPTARG//,/ }
            for param in $params; do
                if [ "$param" == "rw"   ]; then continue; fi
                if [ "$param" == "ro"   ]; then continue; fi
                if [ "$param" == "dev"  ]; then continue; fi
                if [ "$param" == "suid" ]; then continue; fi
                if [ "$param" == "exec" ]; then continue; fi
                if [ "$param" == "auto" ]; then continue; fi
                if [ "$param" == "nodev" ]; then continue; fi
                if [ "$param" == "nosuid" ]; then continue; fi
                if [ "$param" == "noexec" ]; then continue; fi
                if [ "$param" == "noauto" ]; then continue; fi
                if [[ $param == x-systemd.* ]]; then continue; fi
                trans="$trans --$param"
            done
            ;;
        \?)
            echo "Invalid option: -$OPTARG"
            ;;
    esac
done

# exec rclone
trans="$trans $remote $mountpoint"
PATH=$PATH rclone  mount $trans </dev/null >/dev/null 2>/dev/null &

out=`ls -l $dst`
until [ "$out" != 'total 0' ]; do
    out=`ls -l $dst`
    sleep 1
done
```

**Comment by ScuttleSE** [comment]
The line I have in my autofs looks like this

`example@redacted.com	-fstype=fuse,config=/home/scuttle/.config/rclone/rclone.conf,allow-other,default-permissions,allow-non-empty,max-read-ahead=16M,drive-impersonate=example@redacted.com	:rclonefs#remote:`

**Comment by ScuttleSE** [comment]
And `remote:` is a normal google drive remote using a service account json-file

**Comment by ncw** [comment]
OK thank you !

I updated the mount helper script in [the wiki page](https://github.com/ncw/rclone/wiki/rclone-fstab-mount-helper-script) and added a section on autofs.

Can you take a quick look to see if that is OK?

I'm going to declare this solved now and close the issue

Thanks for the debugging.

**Comment by ScuttleSE** [comment]
Cool, looks good!

**Comment by karagian** [comment]
Hi, I tested the wrapper script for use with automount and there may be a way around the restriction for non-empty remote...
Just by replacing the following line with the test in the loop

`until [ "`ls -l $mountpoint`" != 'total 0' ]; do`

with something like that:

`until mountpoint "$mountpoint"; do`

or even 

`until grep -q " ${mountpoint// /\\\\040} " /proc/mounts; do`

seems to work without problems... (the above substitution is necessary for mountpoints that contain spaces in the path)

I think this is a more proper way to check if the filesystem is mounted and resolves the empty remote issue...

**Comment by ivandeex** [comment]
@ncw
I still see problems with systemd automount.
Will investigate, report and open a new issue.
For the time being marking this ticket as "assigned" on me, as a bookmark of sorts.

**Comment by karagian** [comment]
Sorry we don't use systemd so I can't test it with it...
Let me know if I can help

**Comment by ivandeex** [comment]
@ncw

> rclone is blocked on `cmd/mountlib.checkMountEmpty` - it is listing the mountpoint to see if it is empty. This is likely to cause some recursion or other funkiness with autofs.
> If you add the ` --allow-non-empty` flag it will skip that check - perhaps that will make it work!

We should make `checkMountEmpty` a little more clever. It should check `/proc/mounts` before listing directory.
Here is an example of how it looks on my box where I use rclone mounts with systemd automounter:
```
$ grep autofs /proc/mounts
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=37,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=12471 0 0
systemd-1 /mnt/rc/dropbox autofs rw,relatime,fd=89,pgrp=1,timeout=30,minproto=5,maxproto=5,direct,pipe_ino=18165 0 0
systemd-1 /mnt/rc/mailru autofs rw,relatime,fd=90,pgrp=1,timeout=30,minproto=5,maxproto=5,direct,pipe_ino=18168 0 0
```

**Comment by ncw** [comment]
> We should make `checkMountEmpty` a little more clever. It should check `/proc/mounts` before listing directory.

That's very linux centric. Is there a syscall we could issue instead?

> Here is an example of how it looks on my box where I use rclone mounts with systemd automounter:
> 
> ```
> $ grep autofs /proc/mounts
> systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=37,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=12471 0 0
> systemd-1 /mnt/rc/dropbox autofs rw,relatime,fd=89,pgrp=1,timeout=30,minproto=5,maxproto=5,direct,pipe_ino=18165 0 0
> systemd-1 /mnt/rc/mailru autofs rw,relatime,fd=90,pgrp=1,timeout=30,minproto=5,maxproto=5,direct,pipe_ino=18168 0 0
> ```

Nice :-)

**Comment by ivandeex** [comment]
> > We should make `checkMountEmpty` a little more clever. It should check `/proc/mounts` before listing directory.
> 
> That's very linux centric.

Correct. It's a Linux-only solution.

> Is there a syscall we could issue instead?

AFAIK no. Kernels 3.10 - 5.4 expose mount data via procfs only.

**Comment by ncw** [comment]
> > > We should make `checkMountEmpty` a little more clever. It should check `/proc/mounts` before listing directory.
> > 
> > 
> > That's very linux centric.
> 
> Correct. It's a Linux-only solution.
> 
> > Is there a syscall we could issue instead?
> 
> AFAIK no. Kernels 3.10 - 5.4 expose mount data via procfs only.

The classic way of finding out if a directory is a mount point is to `stat` it and check if its device number (`st_dev`) is different from the device number of the parent directory. We do some of this kind of thing in the local backend already to implement the `-x` flag.

However calling stat on the directory may well do the whole recursive thing we are trying to avoid, I'm not sure...

**Comment by ivandeex** [comment]
I can make rclone mount work with `systemd.mount` and `systemd.automount` on Linux using a wrapping shell script.

My dev branch [rclone-for-autofs](https://github.com/ivandeex/rclone/commits/ci-autofs) has two patches against latest rclone `1.56` (of course it can fork using `--daemon`):
1. the parent process can wait for forked child to finish mpunts, using the `--daemon-timeout=30s` flag.
2. neither parent nor child try to list files in the mountpoint, they just parse the Linux API file `/proc/mounts`.
Also I use systemd files from https://gist.github.com/ivandeex/5bb2d26c9e3d01de62285c58cf08f23f:
- /etc/systemd/system/mnt-mysftp.mount
- /etc/systemd/system/mnt-mysftp.automount
- /sbin/mount.rclone

The former file is a shell wrapper invoked by systemd when it establishes a `type=rclone` mount. Systemd passes mount options as `-o option=value` to the wrapper, and expects it to do the mount and exit with zero status.

The path must be fully mounted when the wrapper exits, so the default way `rclone mount --daemon` works (parent exits immediately without waiting for mount) is not enough. The wrapper must ensure that mount is completely established by the time it exits, or systemd will detect failure. I tried two methods to ensure this.

1. I don't use the `--daemon` flag. The wrapper runs `rclone mount ... &` then polls `/proc/mounts` until the mount establishes. This method works fine both with systemd mount and automount.

2. The wrapper just translates `-o opt=val` into rclone flags like `--opt=val` and invokes `rclone mount --daemon --daemon-timeout=30s ....`. This way rclone does all: it forks, the child mounts, the parent waits for the child and then exits.

The 2nd method works fine with systemd mount. After `systemctl start mnt-mysftp.mount` I can access `/mnt/mysftp` without probkems.
But if I stop mount, start automount and try to access `ls /mnt/myfork`, it fails with `rclone: fusermount: /mnt/myfork: ENODEV`. I am investigaing this, trying kludges on the branch, etc...

**Comment by ivandeex** [comment]
@ncw 
I now can identify 5 main reasons why autofs (I use systemd automount variant) failed with rclone:
1. `go-daemon` calls `setsid()` resulting in `pgid` different from main process, so autofs treats it as a "foreign" process (i.e. user of mounted folder, not mounter)
2. rclone accesses to-be mounted folder (lists it etc) being treated as user, so autofs tries to activate mounter again
3. autofs (in particular systemd) expects that mount is complete by the time mounter process exits, but `rclone --daemon` will not wait for _NewFs_` or _Mount_ to complete
4. autofs installs directory hooks in kernel to detect when the dir is accessed, the "hooks" are exposed as another mount of type `autofs`, rclone should take care od this when it checks that directory is empty and "not mounted"
5. after the idle timeout autofs will stop rclone mount. there are two ways to do that: stop the process or call kernel's `umount` on the directory. in the latter case fuse will detect unmounting and trigger rclone which will call `fusermount -u` in the atexit handler. but its own was already unmounted and replaced by autofs hook, so rclone will unmount the hook resulting in autofs unhappy.

see my patches here:
https://github.com/ivandeex/rclone/commits/ci-autofs

@ScuttleSE @karagian 
please give my fix a try (you will also need [this /sbin/mount.rclone wrapper](https://gist.github.com/ivandeex/038ac805293f44e6859c3c876f586c46#file-sbin_mount-rclone__daemon-timeout)):
https://github.com/ivandeex/rclone/releases/tag/v1.56.0-iva08

Can anyone test it on FreeBSD? 🙏

**Comment by ncw** [comment]
@ivandeex wrote:
> I now can identify 5 main reasons why autofs (I use systemd automount variant) failed with rclone:
> 
> 1. `go-daemon` calls `setsid()` resulting in the child's `pgid` being different from the main process, so autofs treats it as a "foreign" process (i.e. user of mounted folder, not mounter).

I thought calling `setsid` was necessary for a proper detach? Should we not be using `--daemon` with autofs?

> 2. rclone accesses to-be-mounted folder (lists it etc) being treated as a user, so autofs tries to activate mounter again. thus FUSE returned ENODEV.

Makes sense.

> 3. autofs (in particular systemd) expects that mount is complete by the time a mounter process exits, but `rclone --daemon` currently does not wait for _NewFs_ or _Mount_ in the child to complete. you could use a wait-loop in wrapper, I added the `--daemon-timeout=30s` flag on Linux.

Makes sense.

> 4. autofs installs directory hooks in the kernel to detect when the dir is accessed. the "hooks" are exposed as _another mount_ of type `autofs`. rclone should take care of this when it checks that directory is empty and "not mounted"

Arg, another corner case! Who would have thought checking a directory is empty was so hard :-)

> 5. autofs after the idle timeout will stop the rclone mount. there are two ways to do that: stop the process or call kernel's `umount` on the directory. in the latter case fuse will detect the unmounting and trigger rclone who in turn will call `fusermount -u` in the atexit handler. but by the time, its own mount had been already unmounted and replaced by the autofs hook, so rclone will just unmount the hook resulting in autofs being angry, a little :-/

He he!
 
> see my patches here:
> https://github.com/ivandeex/rclone/commits/ci-autofs

I took a look at those.

In macOS using `--daemon-timeout` is a setting for the kernel to shut down the FUSE file system if it doesn't respond for that long, which isn't quite the same usage here. I'd suggest using another flag here, then we can set the default to 30s which is probably what you always want (you want to know synchronously that `rclone mount --daemon` failed rather than having to trawl the logs). How about `--fork-timeout` or `--daemon-mount-timeout` something like that?

**Comment by ivandeex** [comment]
> I thought calling `setsid` was necessary for a proper detach?

in classic unix daemonization they used `fork()` with stdout/stderr redirections, terminal detaching and in particular `setsid()`.

neither `go-daemon` nor my rework call `fork()`. we call golang's `StartProcess` with path to self executable, same cli arguments and same environment with one additional var that tells child "you are child, please follow separate instruction flow". we are non-forking daemonizers.

i didn't dig every gory detail. i just mean that we are not classic unix and for some reason setsid works differently for us. in the end, if i use `StartProcess(...{setsid:true,...})` - autofs  linux systemd fails for us, but with `setsid:false` it's fine.

i would like to see how my fix works for others - on *bsd or on linux with classic non-systemd autofs.

i published autofs protocol links above. i noticed the `pgid` trick and thought it's related.

> Should we not be using `--daemon` with autofs?

we must use it.
autofs expects a nonblocking executable which leaves mount after exit.

> In macOS using `--daemon-timeout` is a setting for the kernel to shut down the FUSE file system if it doesn't respond for that long, which isn't quite the same usage here. I'd 

bad name choice. it should have been `--fuse-timeout`.

> then we can set the default to 30s which is probably what you always want (you want to know synchronously that `rclone mount --daemon` failed rather than having to trawl the logs).

> How about `--fork-timeout`

no. we are not forking.

>  `--daemon-mount-timeout` something like that?

maybe

**Comment by ScuttleSE** [comment]
> @ScuttleSE @karagian
> please give my fix a try:

So, I must admit I've been a bit out of the loop here for a while. Right now I have rclone working with autofs, no extra patches, what should I test? :)

**Comment by ivandeex** [comment]
@ScuttleSE
> So, I must admit I've been a bit out of the loop here for a while. Right now I have rclone working with autofs, no extra patches.

I would like to understand your situation.
Did you something special to make it work?
What is your OS, rclone version, autofs config?
If this is not an issue for you, let's close out this ticket. I'd then open another ticket where I'd describe my case.
At any rate the patch fixes my case. 

> what should I test? :)
I'm interested to know that it does not break your case. If you take your working rclone+autofs environment and replace rclone with patched one, will it still work?

**Comment by ScuttleSE** [comment]
Well, back when the bug was originally opened, there were some minor adjustments, but right now it looks like this:

Autofs:
`Music	    -fstype=fuse.rclonefs,config=/home/scuttle/.config/rclone/rclone.conf,allow-other,default-permissions,allow-non-empty,max-read-ahead=16M,uid=1000,gid=1000`

Rclonefs:
```
#!/bin/bash
remote=$1
mountpoint=$2
shift 2

# Process -o parameters
while getopts :o: opts; do
    case $opts in
        o)
            params=${OPTARG//,/ }
            for param in $params; do
                if [ "$param" == "rw"   ]; then continue; fi
                if [ "$param" == "ro"   ]; then continue; fi
                if [ "$param" == "dev"  ]; then continue; fi
                if [ "$param" == "suid" ]; then continue; fi
                if [ "$param" == "exec" ]; then continue; fi
                if [ "$param" == "auto" ]; then continue; fi
                if [ "$param" == "nodev" ]; then continue; fi
                if [ "$param" == "nosuid" ]; then continue; fi
                if [ "$param" == "noexec" ]; then continue; fi
                if [ "$param" == "noauto" ]; then continue; fi
                if [[ $param == x-systemd.* ]]; then continue; fi
                trans="$trans --$param"
            done
            ;;
        \?)
            echo "Invalid option: -$OPTARG"
            ;;
    esac
done

# exec rclone
trans="$trans $remote $mountpoint"
# NOTE: do not try "mount --daemon" here, it does not play well with systemd automount, use '&'!
# NOTE: mount is suid and ignores pre-set PATHs -> specify explicitely
PATH=$PATH rclone mount $trans </dev/null >/dev/null 2>/dev/null &

# WARNING: this will loop forever if remote is actually empty!
until [ "`ls -l $mountpoint`" != 'total 0' ]; do
    sleep 1
done
```

The mount:

```
[music]
type = drive
scope = drive
service_account_file = /home/scuttle/.config/rclone/serviceaccounts01/009.json
team_drive = XXXXXXXXXXXXXXX

```

Tried the new binary, and works fine that one too

```
scuttle@ranma ~ (master)> rclone version
rclone v1.56.0
- os/version: arch rolling (64 bit)
- os/kernel: 5.13.5-arch1-1 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.16.6
- go/linking: dynamic
- go/tags: none

```

**Comment by ivandeex** [comment]
@ScuttleSE
Thanks a lot for checking amd confirming that my patch works in your environment.

I see now: you avoid `--daemon` using the `&` in wrapper.

This solution almost replicates case `1` from above https://github.com/rclone/rclone/issues/3246#issuecomment-886187797:
> I don't use the --daemon flag. The wrapper runs `rclone mount ... &` i.e. backgrounds it, then polls /proc/mounts until the mount establishes. This method works fine both with systemd mount and automount.

Two _minor_ differences:
1. instead of systemd automount you use the autofs daemon,
2. instead of polling `/proc/mounts` for the mount (it's linux-only) you poll mountpoint directory for files (works on **BSD** too but can get stuck if the mounted fs is empty).

@ncw 

The original case reported here can work fine without my patch. I am solving slightly different case. In fact, I do three things:
1. Let autofs (systemd or classic) work with `rclone --daemon` so wrapper can avoid `&` and just `exec rclone`
2. Move the mount readiness check from wrapper into rclone (under flag `--daemon-wait-timeout` or whatever name you aporove), so it can use more robust ProcFS method on Linux with a fallback for BSD
3. Prevent double-unmount of autofs hooks. This problem can be seen only when autofs idle timeout kicks in. @ScuttleSE did not experience this!

With my patch the wrapper is left with only one function: translation of `-o opt=val` mount options into `--opt=val` rclone flags. If I find a way to use cobra for that (e.g. triggered by executable name `/sbin/mount.rclone` expected by classic `/bin/mount -t rclone remote: /mount/point` or `/usr/sbin/rclonefs` expected by `fusermount`), we can get rid of the wrapper completely!

@ncw @ScuttleSE 

I propose to close out this ticket. I can open another one as in fact I'm solving a different problem.

**Comment by karagian** [comment]
Hi and thanks for looking into this @ivandeex , but I must admin that rclone with autofs was working fine, that was not the problem. 
The reason this issue was reopened was these last lines in the wrapper script that will loop forever if the remote is empty! I saw your changes and it should work almost fine. My only objection is when ${mountpoint} contains spaces, in which case that grep will fail. I just propose using ${mountpoint// /\\\\040} istead of just ${mountpoint} there

**Comment by ivandeex** [comment]
@karagian 

> I just propose using ${mountpoint// /\\040} istead of just ${mountpoint} there

Hit `Edit` on the wiki page and add your change. You don't need an issue for that.
https://github.com/rclone/rclone/wiki/rclone-mount-helper-script

**Comment by karagian** [comment]
Oh ok.. can anyone edi that? Isn't it dangerous?

**Comment by ivandeex** [comment]
> Oh ok.. can anyone edit that? Isn't it dangerous?

Yes, any github user.
No, we'll fix after you if anything.

**Comment by karagian** [comment]
ok done :)

**Comment by ivandeex** [comment]
@ncw 

Let's close this ticket.

I will open another PR with ultimate goal to get rid of shell wrappers so rclone like `sshfs` can work natively with `mount` and `autofs`!

```
ln -s /usr/bin/rclone /sbin/mount.rclone
mount -t rclone remote:path /mnt/remote -o vfs-mode=full ...
```

My last open task is to teach cobra translate `-o xxx` into rclone flags...

**Comment by ivandeex** [comment]
@ncw 

### setsid

I found discussion about setsid in the good old `sshfs` project. It works fine as a mount helper (if you symlink it as `/sbin/mount.sshfs`) and fully supports autofs so we could use it as a primer. They wanted to add `setsid()` right after `fork()` in the child - https://github.com/libfuse/sshfs/issues/141. In the end they don't call it by default, only if they need to read a string from stdin - https://github.com/libfuse/sshfs/blob/master/sshfs.c#L1200.

Again, we don't `fork()` like `sshfs`. We run an independent rclone process which does not inherit anything from the "parent" rclone. We just run it with the same cli flags and environment, just let it know its role via a special environment var.

### autofs hooks

Here is how `/proc/mounts` looks when an autofs mountpoint is idle - our path has a `autofs` mount attached:
```
systemd-1 /mnt/rc/jotta autofs rw,relatime,fd=95,pgrp=1,timeout=30,minproto=5,maxproto=5,direct,pipe_ino=20267 0 0
```

Here is how it looks after autofs detects access and invokes rclone to do the mount - two mounts now attached on the path:
```
systemd-1 /mnt/rc/jotta autofs rw,relatime,fd=95,pgrp=1,timeout=30,minproto=5,maxproto=5,direct,pipe_ino=20267 0 0
jotta: /mnt/rc/jotta fuse.rclone rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
```

If we don't access active mount for a long time, autofs triggers idle event and unmounts the rclone mount, but the autofs pseudo-mount stays (the picture returns to the former state above).

The unmount triggers FUSE which returns a error code on a go channel in rclone. rclone exits and calls atexit with `umount(mountPoint)`. This would be correct if rclone finished due to `^C` or `SIGINT` when both real mount and autofs pseudo-mount below it are active. However, if it exits after external unmount, it should not unmount the remaining autofs pseudo-mount (in fact, it's an access hook).

Here is what happens if rclone by mistake unmounts autofs hooks:
```
systemd: Unmounting rclone mount evan...
systemd: mnt-rc-evan.mount: Failed with result 'exit-code'.
systemd: Unmounted rclone mount evan.
systemd: mnt-rc-evan.automount: Got invalid poll event 16 on pipe (fd=49)
systemd: mnt-rc-evan.automount: Failed with result 'resources'.
```

**Comment by ScuttleSE** [comment]
Hm, just wanted to add... none of my idle autofs-mounts shows up in /proc/mounts, they only show up there when they are actually mounted. No matter if it's a rclone-mount or a regular filesystem mount

**Comment by ncw** [comment]
@ivandeex - thank you for the explanation :-)

> Again, we don't fork() like sshfs. We run an independent rclone process which does not inherit anything from the "parent" rclone. We just run it with the same cli flags and environment, just let it know its role via a special environment var.

Yes if we don't fork, we don't need to call setsid - that seems obvious now.

> The unmount triggers FUSE which returns a error code on a go channel in rclone. rclone exits and calls atexit with umount(mountPoint). This would be correct if rclone finished due to ^C or SIGINT when autofs pseudo-mount wasn't there. However, if it exits after external unmount, it should not unmount the remaining autofs pseudo-mount (in fact, it's an access hook).

Gotcha.

**Comment by ivandeex** [comment]
> Hm, just wanted to add... none of my idle autofs-mounts shows up in /proc/mounts, they only show up there when they are actually mounted. No matter if it's a rclone-mount or a regular filesystem mount

Thank you. It explains why you don't hit the unmount bug.
Looks like the pseudo-mount abstraction is used only by systemd automounter while the Autofs daemon uses an alternative kernel mechanism for that. I will mention it when I submit final patch for rclone.
